### DISCUSSION TITLE
Add uuid to mysql

### DISCUSSION THREAD

**Discussion by videni** [diesel-rs/diesel#2715] [Q&A]
I want to implement UUID for mysql using  `UUID_TO_BIN` and  `BIN_TO_UUID`  as  this [article](https://www.mysqltutorial.org/mysql-uuid/) suggests.  how to call the two sql functions please? 

# insert

`UUID_TO_BIN`  to transform `uuid:Uuid` 

# Select

`BIN_TO_UUID `  to transform `uuid:Uuid` 

Here is an [example](https://github.com/forte-music/core/blob/fc9cd6217708b0dd6ae684df3a53276804479c59/src/models/id.rs#L67-L121) writtern for sqlite,  I made a few modifications,  it is not tested, for I don't find any resource about how to specify  `UUID_TO_BIN` and  `BIN_TO_UUID`. 

```
use std::io::Write;
use diesel::backend::Backend;
use diesel::deserialize;
use diesel::deserialize::FromSql;
use diesel::serialize;
use diesel::serialize::Output;
use diesel::sql_types::Binary;
use diesel::sql_types::HasSqlType;
use diesel::types::ToSql;
use diesel::mysql::Mysql;
use uuid::Uuid;

#[derive(Debug, AsExpression, FromSqlRow, Copy, Clone)]
pub struct UUID(Uuid);

impl<DB: Backend + HasSqlType<Binary>> ToSql<Binary, DB> for UUID {
    fn to_sql<W: Write>(&self, out: &mut Output<W, DB>) -> serialize::Result {
        let bytes = self.0.as_bytes();
        <[u8] as ToSql<Binary, DB>>::to_sql(bytes, out)
    }
}

impl FromSql<Binary, Mysql> for UUID {
    fn from_sql(bytes: Option<&<Mysql as Backend>::RawValue>) -> deserialize::Result<Self> {
        let bytes_vec: Vec<u8> = <Vec<u8> as FromSql<Binary, Mysql>>::from_sql(bytes)?;
        Ok(Uuid(Uuid::from_bytes(&bytes_vec)?))
    }
}

```

**Comment by weiznich** [comment]
`UUID_TO_BIN` and `BIN_TO_UUID` can be defined via `sql_function!`/`no_arg_sql_function!` in any third party crate.

**Reply by videni** [reply]
thanks, how can I  call them  in `from_sql ` and `to_sql ` ?  Sorry , I am very new to rust, and diesel.

**Reply by weiznich** [reply]
To be clear: You cannot use `UUID_TO_BIN` or `BIN_TO_UUID` as part of your `FromSql` and `ToSql` implementations. Both implementations run on values already received from the database or on values on rust side before you send something to the database. They cannot invoke any database provided function like `UUID_TO_BIN`/`BIN_TO_UUID`. Those functions needs to be invoked explicitly as part of your query.
